components/l10n/common/locale_util.cc

```diff
@@ -5,78 +5,145 @@
 
 #include "brave/components/l10n/common/locale_util.h"
 
-#include <algorithm>
-#include <vector>
-
-#include "base/strings/string_split.h"
-#include "base/strings/string_util.h"
-#include "base/strings/utf_string_conversions.h"
-#include "ui/base/resource/resource_bundle.h"
+#include "base/no_destructor.h"
+#include "brave/components/l10n/browser/default_locale_util.h"
+#include "brave/components/l10n/common/locale_subtag_info.h"
+#include "brave/components/l10n/common/locale_subtag_parser_util.h"
 
 namespace brave_l10n {
 
 namespace {
 
-const char kDefaultLanguageCode[] = "en";
-const char kDefaultCountryCode[] = "US";
+constexpr char kFallbackLocale[] = "en_US";
+constexpr char kFallbackLanguageCode[] = "en";
+constexpr char kFallbackCountryCode[] = "US";
+
+absl::optional<std::string>& DefaultLocaleForTesting() {
+  static base::NoDestructor<absl::optional<std::string>> locale;
+  return *locale;
+}
 
 }  // namespace
 
-std::string GetLanguageCode(
+ScopedDefaultLocaleForTesting::ScopedDefaultLocaleForTesting(
     const std::string& locale) {
-  const std::vector<std::string> locale_components = base::SplitString(locale,
-      ".", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+  last_locale_ = DefaultLocaleForTesting();
+  DefaultLocaleForTesting() = locale;
+}
+
+ScopedDefaultLocaleForTesting::~ScopedDefaultLocaleForTesting() {
+  DefaultLocaleForTesting() = last_locale_;
+}
+
+std::string GetDefaultLocaleString() {
+  const absl::optional<std::string> locale_for_testing =
+      DefaultLocaleForTesting();
+  if (locale_for_testing) {
+    return *locale_for_testing;
+  }
 
-  if (locale_components.empty()) {
-    return kDefaultLanguageCode;
+  const absl::optional<std::string> default_locale =
+      MaybeGetDefaultLocaleString();
```

Comment:
do we really need to request this locale each time from OS? What about this approach:
```
std::string& CurrentLocale() {
  static base::NoDestructor<std::string> locale(
    MaybeGetDefaultLocaleString().value_or(kFallbackLocale));
  return *locale;
}
```

this way we don't need `DefaultLocaleForTesting()` at all, we can just override `CurrentLocale()` from a scoped test helper.

Reply:
The system locale could change, and for parts of Rewards or Ads we would need to know this.

Reply:
Fixed and merged under "Refactor scoped default locale" commit